pep-0712.rst

```diff
@@ -0,0 +1,326 @@
+third-party libraries. Additionally third-party libraries are able to clue
+type-checkers into their own conversion semantics through added support in
+:func:`~py3.11:typing.dataclass_transform`, meaning users of those libraries
+benefit as well.
+
+Specification
+=============
+
+New ``converter`` parameter
+---------------------------
+
+This specification introduces a new parameter named ``converter`` to the
+:func:`dataclasses.field` function. When an ``__init__`` method is synthesized
+by ``dataclass``-like semantics, if an argument is provided for the field, the
+``dataclass`` object's attribute will be assigned the result of calling the
+converter on the provided argument. If no argument is given and the field was
+constructed with a default value, the ``dataclass`` object's attribute will be
+assigned the result of calling the converter on the provided default.
+
+Adding this parameter also implies the following changes:
+
+* A ``converter`` attribute will be added to :class:`dataclasses.Field`.
+* ``converter`` will be added to the field specifier parameters of arguments
+  provided to :func:`~py3.11:typing.dataclass_transform`'s ``field`` parameter.
+
+Example
+'''''''
+
+.. code-block:: python
+
+  @dataclasses.dataclass
+  class InventoryItem:
+      # `converter` as a type
+      id: int = dataclasses.field(converter=int)
+      skus: tuple[int, ...] = dataclasses.field(converter=tuple[int, ...])
+      # `converter` as a callable
+      names: tuple[str, ...] = dataclasses.field(
+        converter=lambda names: tuple(map(str.lower, names))
+      )
+
+      # The default value is also converted; therefore the following is not a
+      # type error.
+      stock_image_path: pathlib.PurePosixPath = dataclasses.field(
+        converter=pathlib.PurePosixPath, default="assets/unknown.png"
+      )
+
+  item1 = InventoryItem("1", [234, 765], ["PYTHON PLUSHIE", "FLUFFY SNAKE"])
+  # item1 would have the following values:
+  #   id=1
+  #   skus=(234, 765)
+  #   names=('python plushie', 'fluffy snake')
+  #   stock_image_path=pathlib.PurePosixPath("assets/unknown.png")
+
+
+Impact on typing
+----------------
+
+``converter`` arguments are expected to be callable objects which accept a
+single argument and return a type compatible with the field's annotated type.
+The type of the callable's argument is used as the type of the corresponding
+parameter in the synthesized ``__init__`` method.
+
+Type-narrowing the argument type
+''''''''''''''''''''''''''''''''
+
+For the purpose of deducing the type of the argument in the synthesized
+``__init__`` method, the ``converter`` argument's type can be "narrowed" using
+the following rules:
+
+* If the ``converter`` is of type ``Any``, it is assumed to be callable with a
+  single ``Any`` typed-argument.
+* All keyword-only parameters can be ignored.
+* ``**kwargs`` with defaults can be ignored.
+* ``*args`` can be ignored if any parameters precede it. Otherwise if ``*args``
```

Comment:
One big problem I see with this approach is with lambdas. In Python, there's no way to explicitly annotate the parameter types for a lambda, so they must be inferred through bidirectional type inference (i.e. through the context in which the lambda is used). But in the case of the `converter`, there is no available context, so the parameter type will always be `Any` (or `Unknown`, which is pyright's way of designating an "implicit `Any`). I don't see a good solution to this problem other than to recommend that developers never use lambdas for converters if they want proper type checking.

Reply:
Oh yeah good point. I knew lambdas would be troublesome for converters, but forgot to include it here.

I like calling out the danger of lambdas. Would it be too far to encourage type checkers to discourage lambda usage in this section?

Reply:
I'm not sure what it would mean for "type checkers to discourage lambda usage". I don't think it would be appropriate for a type checker to issue an error or a warning here because it is not a type violation to use a lambda. When the strictest type checking modes are enabled, type checkers could point out that a type cannot be inferred. That may be enough to dissuade some developers from using lambdas here.

Reply:
Yeah so my question is if that's appropriate to include in a PEP

Reply:
I think it's sufficient to point out the issue with lambda parameter inference. Different type checkers have their own mechanisms for reporting cases where types cannot be inferred (typically only when stricter type checking modes are enabled). I don't think a PEP should prescribe _how_ or _when_ a type checker should report such a condition.

Reply:
Excellent! Thanks again for your time reading through this 🙏